File: 4/jake_olenick/base_java/Matrix.java
/*========== Matrix.java ==========
  Matrix will hold a 2-d array of doubles and have a default size of 4x4.
  Handles basic matrix maintenence and math.
  Creates transformation matricies for tralation, scale and rotate
=========================*/

import java.io.*;
import java.util.*;

public class Matrix {

    public static final int DEFAULT_SIZE = 4;
    protected double[][] m;

    /*===========Constructors================
      Default constructor creates a 4x4 matrix
      Second constructor creates a 4xN matrix
    */
    public Matrix() {
	m = new double[DEFAULT_SIZE][DEFAULT_SIZE];
    }
    public Matrix(int c) {
	m = new double[DEFAULT_SIZE][c];
    }


    /*========     public double calculateDot() ==========
      Inputs:   int i  
      Returns:  The dot product of a surface normal and
                a view vector 
   
      calculates the dot product of the surface normal to
      triangle points[i], points[i+1], points[i+2] and a 
      view vector (use <0, 0, -1> to start.      


      04/17/12 17:10:13
      jonalf
      ====================*/
    public double calculateDot( int i ) {
	double dot = 0;
	return dot;
    }
    
    /*========  public double[] crossProduct() ==========
      Inputs:   double ax
                double ay
		double az
		double bx
		double by
		double bz  
      Returns:  A double arry of size 3 representing the 
                cross product of <ax, ay, az> and <bx, by, bz>
      
      04/17/12 17:07:07
      jonalf
      ====================*/
    public double[] crossProduct( double ax, double ay, double az, 
				     double bx, double by, double bz ) {

	double[] normal = new double[3];
	return normal;
    }


    /*===========grow================
      Increase the number of columns in a matrix by 10
      You can change the growth factor as you see fit
    */
    public void grow() {

	double[][] n = new double[m.length][m[0].length + 10];
	for (int r=0; r<m.length; r++)
	    for (int c=0; c<m[r].length; c++)
		n[r][c] = m[r][c];
	
	m = n;
    }

    /*======== public void clear() ==========
      Inputs:  
      Returns: 
      Sets every entry in the matrix to 0
      ====================*/
    public void clear() {

	for (int i=0; i<m.length; i++) 
	    for (int j=0; j<m[i].length; j++) 
		m[i][j] = 0;
    }		

    /*===========ident================
      Turns this matrix into the indentity matrix
      You may assume the calling Matrix is square
    */
    public void ident() {
	
	for (int i=0; i<m.length; i++) {
	    for (int j=0; j<m[i].length; j++) {
		
		if (i==j)
		    m[i][j] = 1;
		else
		    m[i][j] = 0;
	    }
	}
    }

    /*===========scalarMult================
      Inputs:  double x
      
      multiply each element of the calling matrix by x
    */
    public void scalarMult( int s ) {

	for (int i=0; i<m.length; i++) 
	    for (int j=0; j<m[i].length; j++) 
		m[i][j] = m[i][j] * s;
    }		

    /*===========matrixMult================
      Multply matrix n by the calling matrix, modify
      the calling matrix to store the result.
      
      eg.
      In the call a.matrixMult(n), n will remain the same
      and a will now be the product of n * a
    */
    public void matrixMult( Matrix n ) {

	double[][] tmp = new double[4][1];

	for (int c=0; c<m[0].length; c++) {
	    for (int r=0; r<4; r++) 
		tmp[r][0] = m[r][c];

	    for (int r=0; r<4; r++)
		m[r][c] = n.m[r][0] * tmp[0][0] +
		    n.m[r][1] * tmp[1][0] +
		    n.m[r][2] * tmp[2][0] +
		    n.m[r][3] * tmp[3][0];
	}
    }
   
    /*===========copy================
      Create and return new matrix that is a duplicate 
      of the calling matrix
    */
    public Matrix copy() {

	Matrix n = new Matrix( m[0].length );
	for (int r=0; r<m.length; r++)
	    for (int c=0; c<m[r].length; c++)
		n.m[r][c] = m[r][c];

	return n;
    }

    /*===========toString================
      Crate a readable String representation of the 
      calling matrix.
    */
    public String toString() {

	String s = "";
	for (int i=0; i<m.length; i++) {
	    for (int j=0; j<m[i].length; j++)
		s = s + m[i][j] + " ";
	    s = s + "\n";
	}
	return s;
    }

    /*===========MakeTranslate================
      Turns the calling matrix into the appropriate
      translation matrix using x, y, and z as the translation
      offsets.
    */
    public void makeTranslate(double x, double y, double z) {

	ident();
	m[0][3] = x;
	m[1][3] = y;
	m[2][3] = z;
    }
    
    /*===========MakeScale================
      Turns the calling matrix into the appropriate scale
      matrix using x, y and z as the scale factors.
    */
    public void makeScale(double x, double y, double z) {

	ident();
	m[0][0] = x;
	m[1][1] = y;
	m[2][2] = z;
    }

    /*=========== MakeRotX ================
      Turns the calling matrix into the appropriate rotation
      matrix using theta as the angle of rotation and X
      as the axis of rotation.
    */
    public void makeRotX(double theta) {
	
	ident();
	m[1][1] = Math.cos( theta );
	m[1][2] = -1 * Math.sin( theta );
	m[2][1] = Math.sin( theta );
	m[2][2] = Math.cos( theta );
    }

    /*=========== MakeRotY ================
      Turns the calling matrix into the appropriate rotation
      matrix using theta as the angle of rotation and Y
      as the axis of rotation.
    */
    public void makeRotY(double theta) {

	ident();
	m[0][0] = Math.cos( theta );
	m[0][2] = -1 * Math.sin( theta );
	m[2][0] = Math.sin( theta );
	m[2][2] = Math.cos( theta );
    }

    /*=========== MakeRotZ ================
      Turns the calling matrix into the appropriate rotation
      matrix using theta as the angle of rotation and axis
      as the axis of rotation.
    */
    public void makeRotZ(double theta) {

	ident();
	m[0][0] = Math.cos( theta );
	m[0][1] = -1 * Math.sin( theta );
	m[1][0] = Math.sin( theta );
	m[1][1] = Math.cos( theta );
    }
    
    /*======== public void makeHermite()) ==========
      Inputs:   
      Returns: 
      
      Turn the calling matrix into a hermite coeficient
      generating matrix

      03/09/12 18:15:58
      jonalf
      ====================*/
    public void makeHermite() {
	
	ident();
	m[0][0] = 2;
	m[0][1] = -2;
	m[0][2] = 1;
	m[0][3] = 1;

	m[1][0] = -3;
	m[1][1] = 3;
	m[1][2] = -2;
	m[1][3] = -1;

	m[3][0] = 1;
	m[3][3] = 0;
    }

    /*======== public void makeBezier()) ==========
      Inputs:   
      Returns: 

      Turn the calling matrix into a bezier coeficient
      generating matrix

      03/09/12 18:15:00
      jonalf
      ====================*/
    public void makeBezier() {
	
	ident();
	m[0][0] = -1;
	m[0][1] = 3;
	m[0][2] = -3;
	m[0][3] = 1;

	m[1][0] = 3;
	m[1][1] = -6;
	m[1][2] = 3;
	m[1][3] = 0;

	m[2][0] = -3;
	m[2][1] = 3;
	m[2][2] = 0;

	m[3][0] = 1;
	m[3][3] = 0;
    }

    /*======== public void generateHermiteCoefs() ==========
      Inputs:  double p1
               double p2
	       double p3
	       double p4 
      Returns: 
      
      Turns the calling matrix into a matrix that provides the 
      coefiecients required to generate a Hermite curve given 
      the values of the 4 parameter coordinates.

      03/09/12 18:17:16
      jonalf
      ====================*/
    public void generateHermiteCoefs(double p1, double p2, 
				     double p3, double p4) {

	Matrix mult = new Matrix(4);

	m[0][0] = p1;
	m[1][0] = p3;
	m[2][0] = p2 - p1;
	m[3][0] = p4 - p3;

	mult.makeHermite();
	matrixMult( mult );
    }

    /*======== public void generateBezierCoefs() ==========
      Inputs:  double p1
               double p2
	       double p3
	       double p4 
      Returns: 
      
      Turns the calling matrix into a matrix that provides the 
      coefiecients required to generate a Bezier curve given 
      the values of the 4 parameter coordinates.

      03/09/12 18:17:16
      jonalf
      ====================*/
    public void generateBezierCoefs(double p1, double p2, 
				    double p3, double p4) {

	Matrix mult = new Matrix(4);

	m[0][0] = p1;
	m[1][0] = p2;
	m[2][0] = p3;
	m[3][0] = p4;

	mult.makeBezier();
	matrixMult( mult );
    } 


}

File: 4/richard_zhan/matrix.py
import math

def make_bezier():
    b = new_matrix()
    ident(b)
    b[0][0] = -1
    b[1][0] = 3
    b[2][0] = -3
    b[3][0] = 1

    b[0][1] = 3
    b[1][1] = -6
    b[2][1] = 3
    b[3][1] = 0

    b[0][2] = -3
    b[1][2] = 3
    b[2][2] = 0
    b[3][2] = 0

    b[0][3] = 1
    b[3][3] = 0
    return b

def make_hermite():
    h = new_matrix()
    ident(h)
    h[0][0] = 2
    h[1][0] = -2
    h[2][0] = 1
    h[3][0] = 1

    h[0][1] = -3
    h[1][1] = 3
    h[2][1] = -2
    h[3][1] = -1

    h[0][2] = 0
    h[1][2] = 0
    h[2][2] = 1
    h[3][2] = 0

    h[0][3] = 1
    h[3][3] = 0
    return h

def generate_curve_coefs( p1, p2, p3, p4, t ):
    coefs = new_matrix(4, 1)

    if t == 'bezier':
        inverse = make_bezier()
        coefs[0][0] = p1
        coefs[0][1] = p2
        coefs[0][2] = p3
        coefs[0][3] = p4
    else:
        inverse = make_hermite()
        coefs[0][0] = p1
        coefs[0][1] = p3
        coefs[0][2] = p2 - p1
        coefs[0][3] = p4 - p3

    matrix_mult( inverse, coefs )
    return coefs

    

def make_translate( x, y, z ):
    t = new_matrix()
    ident(t)
    t[3][0] = x
    t[3][1] = y
    t[3][2] = z
    return t

def make_scale( x, y, z ):
    s = new_matrix()
    ident(s)
    s[0][0] = x
    s[1][1] = y
    s[2][2] = z
    return s
    
def make_rotX( theta ):    
    rx = new_matrix()
    ident( rx )
    rx[1][1] = math.cos( theta )
    rx[2][1] = -1 * math.sin( theta )
    rx[1][2] = math.sin( theta )
    rx[2][2] = math.cos( theta )
    return rx

def make_rotY( theta ):
    ry = new_matrix()
    ident( ry )
    ry[0][0] = math.cos( theta )
    ry[2][0] = -1 * math.sin( theta )
    ry[0][2] = math.sin( theta )
    ry[2][2] = math.cos( theta )
    return ry

def make_rotZ( theta ):
    rz = new_matrix()
    ident( rz )
    rz[0][0] = math.cos( theta )
    rz[1][0] = -1 * math.sin( theta )
    rz[0][1] = math.sin( theta )
    rz[1][1] = math.cos( theta )
    return rz

def new_matrix(rows = 4, cols = 4):
    m = []
    for c in range( cols ):
        m.append( [] )
        for r in range( rows ):
            m[c].append( 0 )
    return m

def print_matrix( matrix ):
    s = ''
    for r in range( len( matrix[0] ) ):
        for c in range( len(matrix) ):
            s+= str(matrix[c][r]) + ' '
        s+= '\n'
    print s

def ident( matrix ):
    for r in range( len( matrix[0] ) ):
        for c in range( len( matrix ) ):
            if r == c:
                matrix[c][r] = 1
            else:
                matrix[c][r] = 0

def scalar_mult( matrix, x ):
    for r in range( len( matrix[0] ) ):
        for c in range( len( matrix ) ):
            matrix[c][r] *= x

#m1 * m2 -> m2
def matrix_mult( m1, m2 ):
    
    t = new_matrix( 4, 1 )

    for c in range( len( m2 ) ):        
        
        for r in range(4):
            t[0][r] = m2[c][r]
            
        for r in range(4):
            m2[c][r] = m1[0][r] * t[0][0] + m1[1][r] * t[0][1] + m1[2][r] * t[0][2] + m1[3][r] * t[0][3]


